src/constraints/absorb.rs

```diff
@@ -0,0 +1,153 @@
+use ark_ec::{ModelParameters, SWModelParameters, TEModelParameters};
+use ark_ff::{Field, PrimeField};
+use ark_r1cs_std::bits::boolean::Boolean;
+use ark_r1cs_std::bits::uint8::UInt8;
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_r1cs_std::fields::{FieldOpsBounds, FieldVar};
+use ark_r1cs_std::groups::curves::short_weierstrass::{
+    AffineVar as SWAffineVar, ProjectiveVar as SWProjectiveVar,
+};
+use ark_r1cs_std::groups::curves::twisted_edwards::AffineVar as TEAffineVar;
+use ark_r1cs_std::ToConstraintFieldGadget;
+use ark_relations::r1cs::SynthesisError;
+use ark_std::vec;
+use ark_std::vec::Vec;
+/// An interface for objects that can be absorbed by a `CryptographicSpongeVar`.
+pub trait AbsorbGadget<F: PrimeField> {
+    /// Converts the object into field elements that can be absorbed by a `CryptographicSpongeVar`.
+    fn to_sponge_field_elements(&self) -> Result<Vec<FpVar<F>>, SynthesisError>;
+
+    /// Specifies the conversion into a list of field elements for a batch.
+    fn batch_to_sponge_field_elements(batch: &[Self]) -> Result<Vec<FpVar<F>>, SynthesisError>
+    where
+        Self: Sized,
+    {
+        let mut output = Vec::new();
+        for absorbable in batch {
+            output.append(&mut absorbable.to_sponge_field_elements()?);
+        }
+
+        Ok(output)
+    }
+}
+
+impl<F: PrimeField> AbsorbGadget<F> for UInt8<F> {
+    fn to_sponge_field_elements(&self) -> Result<Vec<FpVar<F>>, SynthesisError> {
+        vec![self.clone()].to_constraint_field()
+    }
+
+    fn batch_to_sponge_field_elements(batch: &[Self]) -> Result<Vec<FpVar<F>>, SynthesisError> {
+        let mut bytes = UInt8::constant_vec((batch.len() as u64).to_le_bytes().as_ref());
```

Comment:
fwd @Will-Lin4 not exactly sure why, can you explain a little bit? Thanks

Reply:
It's okay to allocate as constant because at circuit-generation time, the length must be statically known (it cannot vary with the variable assignments).

Reply:
Could you add my last comment as a comment in the source code? Thanks!

Reply:
oops. sorry I missed it. Will add to it now!